File: src/backend/whitelist.ts
import type {Request, Response, NextFunction} from "express";
import { execSync } from "child_process";
import { matches } from "ip-matching";
import fs from "fs";
import path from "path";

interface WhitelistConfig {
  enabled: boolean;
  ips: string[];
  macs: string[];
}

const CONFIG_PATH = path.resolve("./whitelist.json");

const DEFAULT_CONFIG: WhitelistConfig = {
  enabled: false,
  ips: ["127.0.0.1", "::1", "::ffff:127.0.0.1"],
  macs: [],
};

export function loadWhitelist(): WhitelistConfig {
  if (!fs.existsSync(CONFIG_PATH)) {
    fs.writeFileSync(CONFIG_PATH, JSON.stringify(DEFAULT_CONFIG, null, 2));
    return DEFAULT_CONFIG;
  }
  try {
    return JSON.parse(fs.readFileSync(CONFIG_PATH, "utf-8"));
  } catch {
    return DEFAULT_CONFIG;
  }
}

export function saveWhitelist(config: WhitelistConfig) {
  fs.writeFileSync(CONFIG_PATH, JSON.stringify(config, null, 2));
}

function arpLookup(ip: string): string | null {
  try {
    const clean = ip.replace("::ffff:", "");
    const output = execSync(`arp -n ${clean} 2>/dev/null || arp -a ${clean} 2>/dev/null`, {
      encoding: "utf-8",
      timeout: 3000,
    });
    const match = output.match(/([0-9a-f]{2}[:-]){5}[0-9a-f]{2}/i);
    return match ? match[0].toLowerCase().replace(/-/g, ":") : null;
  } catch {
    return null;
  }
}

function getClientIp(req: Request): string {
  const forwarded = req.headers["x-forwarded-for"];
  if (typeof forwarded === "string") return forwarded.split(",")[0]?.trim() ?? forwarded.trim();
  return req.socket.remoteAddress ?? "unknown";
}

const LOCALHOST = ["127.0.0.1", "::1", "::ffff:127.0.0.1"];

function isLocalhost(ip: string): boolean {
  return LOCALHOST.some(l => ip === l || ip === `::ffff:${l}`);
}

function checkIp(clientIp: string, allowedIps: string[]): boolean {
  if (allowedIps.length === 0) return true;
  const normalized = clientIp.replace("::ffff:", "");
  return allowedIps.some(pattern => {
    const normalizedPattern = pattern.replace("::ffff:", "");
    try {
      return matches(normalized, normalizedPattern);
    } catch {
      // Fall back to exact match if pattern is invalid
      return clientIp === pattern || clientIp === `::ffff:${pattern}`;
    }
  });
}

function checkMac(clientIp: string, allowedMacs: string[]): boolean {
  if (allowedMacs.length === 0) return true;
  const mac = arpLookup(clientIp);
  if (mac && !allowedMacs.includes(mac)) return false;
  return true;
}

export function deviceWhitelist(req: Request, res: Response, next: NextFunction) {
  const config = loadWhitelist();
  const clientIp = getClientIp(req);

  if (req.path === "/api/whitelist") {
    if (isLocalhost(clientIp)) return next();
    res.status(403).json({ ok: false, message: "Whitelist admin is localhost-only." });
    return;
  }

  if (!config.enabled) {
    if (isLocalhost(clientIp)) return next();
    console.warn(`[whitelist] disabled — blocked non-local IP: ${clientIp}`);
    res.status(403).json({ ok: false, message: "Access restricted to localhost." });
    return;
  }

  if (!checkIp(clientIp, config.ips)) {
    console.warn(`[whitelist] blocked IP: ${clientIp}`);
    res.status(403).json({ ok: false, message: `IP ${clientIp} not whitelisted.` });
    return;
  }

  if (!checkMac(clientIp, config.macs)) {
    const mac = arpLookup(clientIp);
    console.warn(`[whitelist] blocked MAC: ${mac} (IP: ${clientIp})`);
    res.status(403).json({ ok: false, message: "Device not whitelisted." });
    return;
  }

  next();
}

File: src/backend/whitelist.py
import ipaddress
import json
import logging
import re
import subprocess
from pathlib import Path

logger = logging.getLogger(__name__)

CONFIG_PATH = Path("./whitelist.json")
LOCALHOST = {"127.0.0.1", "::1", "::ffff:127.0.0.1"}

DEFAULT_CONFIG: dict = {
    "enabled": False,
    "ips": ["127.0.0.1", "::1", "::ffff:127.0.0.1"],
    "macs": [],
}


def load_whitelist() -> dict:
    if not CONFIG_PATH.exists():
        CONFIG_PATH.write_text(json.dumps(DEFAULT_CONFIG, indent=2))
        return DEFAULT_CONFIG.copy()
    try:
        return json.loads(CONFIG_PATH.read_text())
    except Exception:
        return DEFAULT_CONFIG.copy()


def save_whitelist(config: dict) -> None:
    CONFIG_PATH.write_text(json.dumps(config, indent=2))


def _arp_lookup(ip: str) -> str | None:
    try:
        clean = ip.replace("::ffff:", "")
        result = subprocess.run(["arp", "-n", clean], capture_output=True, text=True, timeout=3)
        if result.returncode != 0:
            result = subprocess.run(["arp", "-a", clean], capture_output=True, text=True, timeout=3)
        match = re.search(r"([0-9a-f]{2}[:-]){5}[0-9a-f]{2}", result.stdout, re.IGNORECASE)
        return match.group(0).lower().replace("-", ":") if match else None
    except Exception:
        return None


def _is_localhost(ip: str) -> bool:
    return ip in LOCALHOST or f"::ffff:{ip}" in LOCALHOST


def _check_ip(client_ip: str, allowed_ips: list[str]) -> bool:
    if not allowed_ips:
        return True
    normalized = client_ip.replace("::ffff:", "")
    for pattern in allowed_ips:
        norm_pattern = pattern.replace("::ffff:", "")
        try:
            if ipaddress.ip_address(normalized) in ipaddress.ip_network(norm_pattern, strict=False):
                return True
        except ValueError:
            if client_ip == pattern or client_ip == f"::ffff:{pattern}":
                return True
    return False


def _check_mac(client_ip: str, allowed_macs: list[str]) -> bool:
    if not allowed_macs:
        return True
    mac = _arp_lookup(client_ip)
    if mac and mac not in allowed_macs:
        return False
    return True


def check_whitelist(client_ip: str, path: str) -> tuple[bool, str]:
    config = load_whitelist()

    if not config["enabled"]:
        if _is_localhost(client_ip):
            return True, ""
        logger.warning("[whitelist] disabled — blocked non-local IP: %s", client_ip)
        return False, "Access restricted to localhost."

    if not _check_ip(client_ip, config["ips"]):
        logger.warning("[whitelist] blocked IP: %s", client_ip)
        return False, f"IP {client_ip} not whitelisted."

    if not _check_mac(client_ip, config["macs"]):
        mac = _arp_lookup(client_ip)
        logger.warning("[whitelist] blocked MAC: %s (IP: %s)", mac, client_ip)
        return False, "Device not whitelisted."

    return True, ""
